File: apps/ios/SignalASI/VoiceInteractionCoordinator.swift
import Foundation

typealias VoiceElapsedClock = () -> Int64
typealias VoiceInteractionObserver = (VoiceInteractionState) throws -> Void

final class VoiceInteractionCoordinator {
  private let elapsedClock: VoiceElapsedClock
  private let sessionIdFactory: () -> String
  private let lock = NSLock()
  private var observers: [String: VoiceInteractionObserver] = [:]
  private var emittedCommandKeys: Set<String> = []
  private var emittedCommandOrder: [String] = []
  private var currentConfig: VoiceSessionConfig?
  private var state = VoiceInteractionState()

  init(
    elapsedClock: @escaping VoiceElapsedClock = VoiceInteractionCoordinator.defaultElapsedClock,
    sessionIdFactory: @escaping () -> String = { UUID().uuidString }
  ) {
    self.elapsedClock = elapsedClock
    self.sessionIdFactory = sessionIdFactory
  }

  func snapshot() -> VoiceInteractionState {
    locked { state }
  }

  func config() -> VoiceSessionConfig? {
    locked { currentConfig }
  }

  func observe(_ observer: @escaping VoiceInteractionObserver) -> String {
    let observerId = UUID().uuidString
    let current = locked { () -> VoiceInteractionState in
      observers[observerId] = observer
      return state
    }
    try? observer(current)
    return observerId
  }

  func removeObserver(_ observerId: String) {
    locked {
      observers.removeValue(forKey: observerId)
    }
  }

  func begin(config: VoiceSessionConfig) -> VoiceInteractionTransition {
    let result = locked { () -> (VoiceInteractionTransition, VoiceInteractionState?, [VoiceInteractionObserver]) in
      let previous = state
      if !state.phase.isTerminal && state.phase != .idle {
        return (VoiceInteractionTransition(previous: previous, current: previous, accepted: false), nil, [])
      }
      let trimmedSessionId = config.requestedSessionId.trimmingCharacters(in: .whitespacesAndNewlines)
      let sessionId = trimmedSessionId.isEmpty ? sessionIdFactory() : trimmedSessionId
      if previous.sessionId == sessionId && previous.phase.isTerminal {
        return (VoiceInteractionTransition(previous: previous, current: previous, accepted: false), nil, [])
      }
      let now = max(0, elapsedClock())
      currentConfig = VoiceSessionConfig(
        requestedSessionId: sessionId,
        source: config.source,
        language: config.language,
        targetId: config.targetId,
        routingMode: config.routingMode,
        speakReplies: config.speakReplies,
        continueInBackground: config.continueInBackground
      )
      state = VoiceInteractionState(
        sessionId: sessionId,
        phase: .preparing,
        revision: previous.revision + 1,
        createdAtElapsedNs: now,
        updatedAtElapsedNs: now
      )
      trimCommandLedger()
      return (
        VoiceInteractionTransition(previous: previous, current: state),
        state,
        Array(observers.values)
      )
    }
    notify(result.1, observers: result.2)
    return result.0
  }

  func dispatch(_ event: VoiceInteractionEvent) -> VoiceInteractionTransition {
    let result = locked { () -> (VoiceInteractionTransition, VoiceInteractionState?, [VoiceInteractionObserver]) in
      let previous = state
      if event.sessionId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
          event.sessionId != previous.sessionId ||
          (previous.phase.isTerminal && !event.isTranscriptCorrection) {
        return (VoiceInteractionTransition(previous: previous, current: previous, accepted: false), nil, [])
      }
      let reduced = reduce(previous: previous, event: event)
      if reduced.state == previous && reduced.commands.isEmpty {
        return (VoiceInteractionTransition(previous: previous, current: previous, accepted: false), nil, [])
      }
      let now = max(previous.updatedAtElapsedNs, elapsedClock())
      state = reduced.state
      state.revision = previous.revision + 1
      state.updatedAtElapsedNs = max(0, now)
      let commands = reduced.commands.filter { acceptCommandKey($0.idempotencyKey) }
      trimCommandLedger()
      return (
        VoiceInteractionTransition(previous: previous, current: state, commands: commands),
        state,
        Array(observers.values)
      )
    }
    notify(result.1, observers: result.2)
    return result.0
  }

  func cancel(reasonCode: String = "user_cancelled") -> VoiceInteractionTransition {
    let current = snapshot()
    guard !current.sessionId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      return VoiceInteractionTransition(previous: current, current: current, accepted: false)
    }
    return dispatch(.cancelled(sessionId: current.sessionId, reasonCode: reasonCode))
  }

  func result() -> VoiceSessionResult? {
    locked {
      guard state.phase.isTerminal, !state.sessionId.isEmpty else { return nil }
      return VoiceSessionResult(
        sessionId: state.sessionId,
        completed: state.phase == .completed,
        cancelled: state.phase == .cancelled,
        route: state.route,
        failure: state.failure,
        finalTranscriptRevision: state.finalTranscriptRevision
      )
    }
  }

  private func reduce(
    previous: VoiceInteractionState,
    event: VoiceInteractionEvent
  ) -> (state: VoiceInteractionState, commands: [VoiceInteractionCommand]) {
    switch event {
    case .capturePrepared(sessionId: _):
      guard previous.phase == .preparing else { return (previous, []) }
      var next = previous
      next.phase = .listening
      return (next, [])

    case .audioLevel(sessionId: _, rms: _):
      return (previous, [])

    case .speechStarted(sessionId: _, atElapsedNs: _):
      guard previous.phase == .preparing || previous.phase == .listening else { return (previous, []) }
      var next = previous
      next.phase = .listening
      return (next, [])

    case .speechEnded(sessionId: _, atElapsedNs: _):
      guard previous.phase == .listening else { return (previous, []) }
      var next = previous
      next.phase = .endpointing
      return (next, [])

    case .finalizationStarted(sessionId: _):
      guard previous.phase == .listening || previous.phase == .endpointing else { return (previous, []) }
      var next = previous
      next.phase = .finalizingASR
      return (next, [])

    case let .transcriptPartial(sessionId: _, value: value):
      guard transcriptUpdatesAllowed(previous.phase) else { return (previous, []) }
      var next = previous
      next.partialText = value.text
      next.asrProvider = nonBlank(value.provider) ?? previous.asrProvider
      next.modelProfileId = nonBlank(value.modelProfileId) ?? previous.modelProfileId
      return (next, [])

    case let .transcriptStable(sessionId: _, value: value):
      guard transcriptUpdatesAllowed(previous.phase) else { return (previous, []) }
      var next = previous
      next.stableText = value.text
      next.asrProvider = nonBlank(value.provider) ?? previous.asrProvider
      next.modelProfileId = nonBlank(value.modelProfileId) ?? previous.modelProfileId
      return (next, [])

    case let .transcriptFinal(sessionId: _, value: value):
      let normalized = value.text.trimmingCharacters(in: .whitespacesAndNewlines)
      guard !normalized.isEmpty else { return (previous, []) }
      if previous.finalText == nil && transcriptUpdatesAllowed(previous.phase) {
        var finalValue = value
        finalValue.text = normalized
        var next = previous
        next.phase = .routing
        next.partialText = ""
        next.stableText = normalized
        next.finalText = normalized
        next.finalTranscriptRevision = value.revision
        next.asrProvider = nonBlank(value.provider) ?? previous.asrProvider
        next.modelProfileId = nonBlank(value.modelProfileId) ?? previous.modelProfileId
        let key = "\(previous.sessionId):route:\(value.revision)"
        return (
          next,
          [.routeFinalTranscript(sessionId: previous.sessionId, transcript: finalValue, idempotencyKey: key)]
        )
      }
      if previous.finalText != normalized {
        var next = previous
        next.correctedText = normalized
        return (next, [])
      }
      return (previous, [])

    case let .transcriptCorrected(sessionId: _, original: _, corrected: corrected):
      let text = corrected.text.trimmingCharacters(in: .whitespacesAndNewlines)
      guard !text.isEmpty, text != previous.correctedText else { return (previous, []) }
      var next = previous
      next.correctedText = text
      return (next, [])

    case let .routeSelected(sessionId: _, decision: decision):
      guard previous.phase == .routing else { return (previous, []) }
      var next = previous
      next.route = decision
      switch decision.kind {
      case .localAction:
        next.phase = .executingLocalAction
      case .cloudModel:
        next.phase = .waitingModelFirstToken
      case .remoteAgent:
        next.phase = .startingAgent
      }
      return (next, [])

    case .localActionCompleted(sessionId: _):
      guard previous.phase == .executingLocalAction else { return (previous, []) }
      var next = previous
      next.phase = .completed
      next.canInterrupt = false
      return (next, [])

    case .modelDelta(sessionId: _, text: _):
      guard previous.phase == .waitingModelFirstToken || previous.phase == .streamingModelText else {
        return (previous, [])
      }
      var next = previous
      next.phase = .streamingModelText
      return (next, [])

    case let .agentAccepted(sessionId: _, runId: runId),
         let .agentProgress(sessionId: _, runId: runId):
      guard previous.phase == .startingAgent || previous.phase == .agentRunning else {
        return (previous, [])
      }
      var next = previous
      next.phase = .agentRunning
      next.agentRunId = nonBlank(runId) ?? previous.agentRunId
      return (next, [])

    case .playbackStarted(sessionId: _, utteranceId: _):
      guard previous.phase == .waitingModelFirstToken ||
          previous.phase == .streamingModelText ||
          previous.phase == .agentRunning else {
        return (previous, [])
      }
      var next = previous
      next.phase = .playingTTS
      return (next, [])

    case .completed(sessionId: _):
      var next = previous
      next.phase = .completed
      next.canInterrupt = false
      return (next, [])

    case let .cancelled(sessionId: _, reasonCode: reasonCode):
      var next = previous
      next.phase = .cancelled
      next.canInterrupt = false
      return (
        next,
        [.cancelLegacyWork(
          sessionId: previous.sessionId,
          reasonCode: reasonCode,
          idempotencyKey: "\(previous.sessionId):cancel"
        )]
      )

    case let .failed(sessionId: _, failure: failure):
      var next = previous
      next.phase = .failed
      next.canInterrupt = false
      next.failure = failure
      return (next, [])
    }
  }

  private func notify(
    _ state: VoiceInteractionState?,
    observers: [VoiceInteractionObserver]
  ) {
    guard let state = state else { return }
    observers.forEach { observer in
      try? observer(state)
    }
  }

  private func acceptCommandKey(_ key: String) -> Bool {
    guard emittedCommandKeys.insert(key).inserted else { return false }
    emittedCommandOrder.append(key)
    return true
  }

  private func trimCommandLedger() {
    guard emittedCommandOrder.count > 2_048 else { return }
    let expired = emittedCommandOrder.prefix(512)
    expired.forEach { emittedCommandKeys.remove($0) }
    emittedCommandOrder.removeFirst(min(512, emittedCommandOrder.count))
  }

  private func locked<T>(_ action: () -> T) -> T {
    lock.lock()
    defer { lock.unlock() }
    return action()
  }

  private func transcriptUpdatesAllowed(_ phase: VoiceInteractionPhase) -> Bool {
    phase == .listening || phase == .endpointing || phase == .finalizingASR
  }

  private func nonBlank(_ value: String) -> String? {
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? nil : trimmed
  }

  private static func defaultElapsedClock() -> Int64 {
    Int64(ProcessInfo.processInfo.systemUptime * 1_000_000_000)
  }
}

enum VoiceInteractionCoordinatorRegistry {
  static let coordinator = VoiceInteractionCoordinator()
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/voice/VoiceInteractionCoordinator.kt
package com.galaxyssi.chat.voice

import android.os.SystemClock
import java.util.UUID

fun interface VoiceElapsedClock {
    fun nowNanos(): Long
}

class VoiceInteractionCoordinator(
    private val elapsedClock: VoiceElapsedClock = VoiceElapsedClock(SystemClock::elapsedRealtimeNanos),
    private val sessionIdFactory: () -> String = { UUID.randomUUID().toString() }
) {
    private val observers = linkedMapOf<String, (VoiceInteractionState) -> Unit>()
    private val emittedCommandKeys = LinkedHashSet<String>()
    private var currentConfig: VoiceSessionConfig? = null
    private var state = VoiceInteractionState()

    @Synchronized
    fun snapshot(): VoiceInteractionState = state

    @Synchronized
    fun config(): VoiceSessionConfig? = currentConfig

    fun observe(observer: (VoiceInteractionState) -> Unit): String {
        val observerId = UUID.randomUUID().toString()
        val snapshot = synchronized(this) {
            observers[observerId] = observer
            state
        }
        runCatching { observer(snapshot) }
        return observerId
    }

    @Synchronized
    fun removeObserver(observerId: String) {
        observers.remove(observerId)
    }

    fun begin(config: VoiceSessionConfig): VoiceInteractionTransition {
        var notification: Pair<VoiceInteractionState, List<(VoiceInteractionState) -> Unit>>? = null
        val transition = synchronized(this) {
            val previous = state
            if (!state.phase.isTerminal && state.phase != VoiceInteractionPhase.IDLE) {
                return@synchronized VoiceInteractionTransition(previous, previous, accepted = false)
            }
            val sessionId = config.requestedSessionId.trim().ifBlank(sessionIdFactory)
            if (previous.sessionId == sessionId && previous.phase.isTerminal) {
                return@synchronized VoiceInteractionTransition(previous, previous, accepted = false)
            }
            val now = elapsedClock.nowNanos().coerceAtLeast(0L)
            currentConfig = config.copy(requestedSessionId = sessionId)
            state = VoiceInteractionState(
                sessionId = sessionId,
                phase = VoiceInteractionPhase.PREPARING,
                revision = previous.revision + 1L,
                createdAtElapsedNs = now,
                updatedAtElapsedNs = now
            )
            trimCommandLedger()
            notification = state to observers.values.toList()
            VoiceInteractionTransition(previous, state)
        }
        notification?.let { update ->
            update.second.forEach { observer -> runCatching { observer(update.first) } }
        }
        return transition
    }

    fun dispatch(event: VoiceInteractionEvent): VoiceInteractionTransition {
        var notification: Pair<VoiceInteractionState, List<(VoiceInteractionState) -> Unit>>? = null
        val transition = synchronized(this) {
            val previous = state
            if (event.sessionId.isBlank() || event.sessionId != previous.sessionId ||
                (previous.phase.isTerminal && event !is VoiceInteractionEvent.TranscriptCorrected)
            ) {
                return@synchronized VoiceInteractionTransition(previous, previous, accepted = false)
            }
            val reduced = reduce(previous, event)
            if (reduced.first == previous && reduced.second.isEmpty()) {
                return@synchronized VoiceInteractionTransition(previous, previous, accepted = false)
            }
            state = reduced.first.copy(
                revision = previous.revision + 1L,
                updatedAtElapsedNs = elapsedClock.nowNanos().coerceAtLeast(previous.updatedAtElapsedNs)
            )
            val commands = reduced.second.filter { emittedCommandKeys.add(it.idempotencyKey) }
            trimCommandLedger()
            notification = state to observers.values.toList()
            VoiceInteractionTransition(previous, state, commands)
        }
        notification?.let { update ->
            update.second.forEach { observer -> runCatching { observer(update.first) } }
        }
        return transition
    }

    fun cancel(reasonCode: String = "user_cancelled"): VoiceInteractionTransition {
        val current = snapshot()
        if (current.sessionId.isBlank()) {
            return VoiceInteractionTransition(current, current, accepted = false)
        }
        return dispatch(VoiceInteractionEvent.Cancelled(current.sessionId, reasonCode))
    }

    @Synchronized
    fun result(): VoiceSessionResult? {
        if (!state.phase.isTerminal || state.sessionId.isBlank()) return null
        return VoiceSessionResult(
            sessionId = state.sessionId,
            completed = state.phase == VoiceInteractionPhase.COMPLETED,
            cancelled = state.phase == VoiceInteractionPhase.CANCELLED,
            route = state.route,
            failure = state.failure,
            finalTranscriptRevision = state.finalTranscriptRevision
        )
    }

    private fun reduce(
        previous: VoiceInteractionState,
        event: VoiceInteractionEvent
    ): Pair<VoiceInteractionState, List<VoiceInteractionCommand>> = when (event) {
        is VoiceInteractionEvent.CapturePrepared -> when (previous.phase) {
            VoiceInteractionPhase.PREPARING -> previous.copy(phase = VoiceInteractionPhase.LISTENING) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.AudioLevel -> previous to emptyList()
        is VoiceInteractionEvent.SpeechStarted -> when (previous.phase) {
            VoiceInteractionPhase.PREPARING,
            VoiceInteractionPhase.LISTENING -> previous.copy(phase = VoiceInteractionPhase.LISTENING) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.SpeechEnded -> when (previous.phase) {
            VoiceInteractionPhase.LISTENING -> previous.copy(phase = VoiceInteractionPhase.ENDPOINTING) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.FinalizationStarted -> when (previous.phase) {
            VoiceInteractionPhase.LISTENING,
            VoiceInteractionPhase.ENDPOINTING -> previous.copy(phase = VoiceInteractionPhase.FINALIZING_ASR) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.TranscriptPartial -> when (previous.phase) {
            VoiceInteractionPhase.LISTENING,
            VoiceInteractionPhase.ENDPOINTING,
            VoiceInteractionPhase.FINALIZING_ASR -> previous.copy(
                partialText = event.value.text,
                asrProvider = event.value.provider.takeIf(String::isNotBlank) ?: previous.asrProvider,
                modelProfileId = event.value.modelProfileId.takeIf(String::isNotBlank)
                    ?: previous.modelProfileId
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.TranscriptStable -> when (previous.phase) {
            VoiceInteractionPhase.LISTENING,
            VoiceInteractionPhase.ENDPOINTING,
            VoiceInteractionPhase.FINALIZING_ASR -> previous.copy(
                stableText = event.value.text,
                asrProvider = event.value.provider.takeIf(String::isNotBlank) ?: previous.asrProvider,
                modelProfileId = event.value.modelProfileId.takeIf(String::isNotBlank)
                    ?: previous.modelProfileId
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.TranscriptFinal -> {
            val normalized = event.value.text.trim()
            when {
                normalized.isBlank() -> previous to emptyList()
                previous.finalText == null && previous.phase in setOf(
                    VoiceInteractionPhase.LISTENING,
                    VoiceInteractionPhase.ENDPOINTING,
                    VoiceInteractionPhase.FINALIZING_ASR
                ) -> {
                    val key = "${previous.sessionId}:route:${event.value.revision}"
                    previous.copy(
                        phase = VoiceInteractionPhase.ROUTING,
                        partialText = "",
                        stableText = normalized,
                        finalText = normalized,
                        finalTranscriptRevision = event.value.revision,
                        asrProvider = event.value.provider.takeIf(String::isNotBlank) ?: previous.asrProvider,
                        modelProfileId = event.value.modelProfileId.takeIf(String::isNotBlank)
                            ?: previous.modelProfileId
                    ) to listOf(
                        VoiceInteractionCommand.RouteFinalTranscript(
                            previous.sessionId,
                            event.value.copy(text = normalized),
                            key
                        )
                    )
                }
                previous.finalText != normalized -> previous.copy(correctedText = normalized) to emptyList()
                else -> previous to emptyList()
            }
        }
        is VoiceInteractionEvent.TranscriptCorrected -> {
            val corrected = event.corrected.text.trim()
            if (corrected.isBlank() || corrected == previous.correctedText) {
                previous to emptyList()
            } else {
                previous.copy(correctedText = corrected) to emptyList()
            }
        }
        is VoiceInteractionEvent.RouteSelected -> when (previous.phase) {
            VoiceInteractionPhase.ROUTING -> previous.copy(
                route = event.decision,
                phase = when (event.decision.kind) {
                    VoiceRouteKind.LOCAL_ACTION -> VoiceInteractionPhase.EXECUTING_LOCAL_ACTION
                    VoiceRouteKind.CLOUD_MODEL -> VoiceInteractionPhase.WAITING_MODEL_FIRST_TOKEN
                    VoiceRouteKind.REMOTE_AGENT -> VoiceInteractionPhase.STARTING_AGENT
                }
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.LocalActionCompleted -> when (previous.phase) {
            VoiceInteractionPhase.EXECUTING_LOCAL_ACTION -> previous.copy(
                phase = VoiceInteractionPhase.COMPLETED,
                canInterrupt = false
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.ModelDelta -> when (previous.phase) {
            VoiceInteractionPhase.WAITING_MODEL_FIRST_TOKEN,
            VoiceInteractionPhase.STREAMING_MODEL_TEXT -> previous.copy(
                phase = VoiceInteractionPhase.STREAMING_MODEL_TEXT
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.AgentRunCreated -> when (previous.phase) {
            VoiceInteractionPhase.STARTING_AGENT -> previous.copy(
                phase = VoiceInteractionPhase.COMPLETED,
                agentRunId = event.runId.takeIf(String::isNotBlank) ?: previous.agentRunId,
                canInterrupt = false
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.AgentAccepted -> when (previous.phase) {
            VoiceInteractionPhase.STARTING_AGENT,
            VoiceInteractionPhase.AGENT_RUNNING -> previous.copy(
                phase = VoiceInteractionPhase.AGENT_RUNNING,
                agentRunId = event.runId.takeIf(String::isNotBlank) ?: previous.agentRunId
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.AgentProgress -> when (previous.phase) {
            VoiceInteractionPhase.STARTING_AGENT,
            VoiceInteractionPhase.AGENT_RUNNING -> previous.copy(
                phase = VoiceInteractionPhase.AGENT_RUNNING,
                agentRunId = event.runId.takeIf(String::isNotBlank) ?: previous.agentRunId
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.PlaybackStarted -> when (previous.phase) {
            VoiceInteractionPhase.WAITING_MODEL_FIRST_TOKEN,
            VoiceInteractionPhase.STREAMING_MODEL_TEXT,
            VoiceInteractionPhase.AGENT_RUNNING -> previous.copy(
                phase = VoiceInteractionPhase.PLAYING_TTS
            ) to emptyList()
            else -> previous to emptyList()
        }
        is VoiceInteractionEvent.Completed -> previous.copy(
            phase = VoiceInteractionPhase.COMPLETED,
            canInterrupt = false
        ) to emptyList()
        is VoiceInteractionEvent.Cancelled -> previous.copy(
            phase = VoiceInteractionPhase.CANCELLED,
            canInterrupt = false
        ) to listOf(
            VoiceInteractionCommand.CancelLegacyWork(
                previous.sessionId,
                event.reasonCode,
                "${previous.sessionId}:cancel"
            )
        )
        is VoiceInteractionEvent.Failed -> previous.copy(
            phase = VoiceInteractionPhase.FAILED,
            canInterrupt = false,
            failure = event.failure
        ) to emptyList()
    }

    @Synchronized
    private fun trimCommandLedger() {
        if (emittedCommandKeys.size <= 2_048) return
        val iterator = emittedCommandKeys.iterator()
        repeat(512) {
            if (iterator.hasNext()) {
                iterator.next()
                iterator.remove()
            }
        }
    }
}

object VoiceInteractionCoordinatorRegistry {
    val coordinator: VoiceInteractionCoordinator by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        VoiceInteractionCoordinator()
    }
}
